File: SecureChat.Client/Storage/LocalHistoryDB.cs
using System;
using System.Collections.Generic;
using System.IO;
using Microsoft.Data.Sqlite;

namespace SecureChat.Client.Storage
{
    public class LocalHistoryDB
    {
        private static readonly string DbDir = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.UserProfile), ".securechat");
        private readonly string _dbPath;
        private SqliteConnection? _connection;

        public LocalHistoryDB(string username)
        {
            _dbPath = Path.Combine(DbDir, $"{username}_history.db");
        }

        /// <summary>
        /// Khởi tạo thư mục và kết nối SQLite, tạo bảng nếu chưa có.
        /// </summary>
        public void Init()
        {
            try
            {
                if (!Directory.Exists(DbDir))
                {
                    Directory.CreateDirectory(DbDir);
                }

                _connection = new SqliteConnection($"Data Source={_dbPath}");
                _connection.Open();
                CreateTables();
                Console.WriteLine($"[LocalDB] SQLite được kết nối thành công: {_dbPath}");
            }
            catch (Exception ex)
            {
                Console.WriteLine($"[LocalDB] Lỗi khởi tạo SQLite: {ex.Message}");
            }
        }

        private void CreateTables()
        {
            string sql = @"
                CREATE TABLE IF NOT EXISTS local_messages (
                    id           INTEGER PRIMARY KEY AUTOINCREMENT,
                    room_id      INTEGER NOT NULL,
                    sender_id    INTEGER NOT NULL,
                    sender_name  TEXT    NOT NULL,
                    content      TEXT    NOT NULL,
                    message_type TEXT    DEFAULT 'TEXT',
                    file_name    TEXT,
                    timestamp    INTEGER NOT NULL,
                    is_sent      INTEGER DEFAULT 1
                );
                CREATE INDEX IF NOT EXISTS idx_room_time ON local_messages(room_id, timestamp);
            ";

            using (var cmd = new SqliteCommand(sql, _connection))
            {
                cmd.ExecuteNonQuery();
            }
        }

        /// <summary>
        /// Lưu tin nhắn đã giải mã vào SQLite.
        /// </summary>
        public void SaveMessage(int roomId, int senderId, string senderName, string content, string type, string fileName, long timestamp, bool isSent)
        {
            string sql = @"
                INSERT INTO local_messages (room_id, sender_id, sender_name, content, message_type, file_name, timestamp, is_sent)
                VALUES ($roomId, $senderId, $senderName, $content, $type, $fileName, $timestamp, $isSent)
            ";

            try
            {
                using (var cmd = new SqliteCommand(sql, _connection))
                {
                    cmd.Parameters.AddWithValue("$roomId", roomId);
                    cmd.Parameters.AddWithValue("$senderId", senderId);
                    cmd.Parameters.AddWithValue("$senderName", senderName);
                    cmd.Parameters.AddWithValue("$content", content);
                    cmd.Parameters.AddWithValue("$type", type);
                    cmd.Parameters.AddWithValue("$fileName", string.IsNullOrEmpty(fileName) ? DBNull.Value : fileName);
                    cmd.Parameters.AddWithValue("$timestamp", timestamp);
                    cmd.Parameters.AddWithValue("$isSent", isSent ? 1 : 0);
                    cmd.ExecuteNonQuery();
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine($"[LocalDB] Lỗi lưu message cục bộ: {ex.Message}");
            }
        }

        /// <summary>
        /// Lấy lịch sử tin nhắn của một phòng (sắp xếp tăng dần theo thời gian).
        /// </summary>
        public List<object[]> GetHistory(int roomId, int limit)
        {
            var results = new List<object[]>();
            string sql = @"
                SELECT room_id, sender_id, sender_name, content, message_type, file_name, timestamp, is_sent
                FROM local_messages
                WHERE room_id = $roomId
                ORDER BY timestamp DESC
                LIMIT $limit
            ";

            try
            {
                using (var cmd = new SqliteCommand(sql, _connection))
                {
                    cmd.Parameters.AddWithValue("$roomId", roomId);
                    cmd.Parameters.AddWithValue("$limit", limit);

                    using (var reader = cmd.ExecuteReader())
                    {
                        while (reader.Read())
                        {
                            results.Add(new object[]
                            {
                                reader.GetInt32(0),
                                reader.GetInt32(1),
                                reader.GetString(2),
                                reader.GetString(3),
                                reader.GetString(4),
                                reader.IsDBNull(5) ? "" : reader.GetString(5),
                                reader.GetInt64(6),
                                reader.GetInt32(7) == 1
                            });
                        }
                    }
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine($"[LocalDB] Lỗi đọc lịch sử: {ex.Message}");
            }

            // Đảo ngược danh sách để hiển thị theo thời gian tăng dần
            results.Reverse();
            return results;
        }

        /// <summary>
        /// Tìm kiếm tin nhắn theo từ khóa.
        /// </summary>
        public List<object[]> SearchMessages(string keyword)
        {
            var results = new List<object[]>();
            string sql = @"
                SELECT room_id, sender_name, content, timestamp
                FROM local_messages
                WHERE content LIKE $keyword
                ORDER BY timestamp DESC
                LIMIT 50
            ";

            try
            {
                using (var cmd = new SqliteCommand(sql, _connection))
                {
                    cmd.Parameters.AddWithValue("$keyword", $"%{keyword}%");
                    using (var reader = cmd.ExecuteReader())
                    {
                        while (reader.Read())
                        {
                            results.Add(new object[]
                            {
                                reader.GetInt32(0),
                                reader.GetString(1),
                                reader.GetString(2),
                                reader.GetInt64(3)
                            });
                        }
                    }
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine($"[LocalDB] Lỗi tìm kiếm: {ex.Message}");
            }
            return results;
        }

        /// <summary>
        /// Xóa lịch sử tin nhắn phòng.
        /// </summary>
        public void ClearRoomHistory(int roomId)
        {
            string sql = "DELETE FROM local_messages WHERE room_id = $roomId";
            try
            {
                using (var cmd = new SqliteCommand(sql, _connection))
                {
                    cmd.Parameters.AddWithValue("$roomId", roomId);
                    cmd.ExecuteNonQuery();
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine($"[LocalDB] Lỗi xoá lịch sử phòng {roomId}: {ex.Message}");
            }
        }

        public void Close()
        {
            try
            {
                _connection?.Close();
            }
            catch { }
        }
    }
}

File: client/src/main/java/com/securechat/client/storage/LocalHistoryDB.java
package com.securechat.client.storage;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.sql.*;
import java.util.ArrayList;
import java.util.List;

/**
 * LocalHistoryDB – Lưu lịch sử chat local bằng SQLite.
 *
 * Database được lưu tại: {user.home}/.securechat/history.db
 * Mỗi user có file DB riêng (theo username).
 */
public class LocalHistoryDB {

    private static final Logger log = LoggerFactory.getLogger(LocalHistoryDB.class);

    private static final String DB_DIR    = System.getProperty("user.home") + "/.securechat/";
    private String dbPath;
    private Connection connection;

    // ────────────────────────────────────────────────────────────

    public LocalHistoryDB(String username) {
        this.dbPath = DB_DIR + username + "_history.db";
    }

    /**
     * Khởi tạo kết nối và tạo bảng nếu chưa có.
     */
    public void init() {
        try {
            // Tạo thư mục nếu chưa có
            java.io.File dir = new java.io.File(DB_DIR);
            if (!dir.exists()) dir.mkdirs();

            Class.forName("org.sqlite.JDBC");
            connection = DriverManager.getConnection("jdbc:sqlite:" + dbPath);
            createTables();
            log.info("SQLite DB khởi tạo thành công: {}", dbPath);
        } catch (Exception e) {
            log.error("Lỗi khởi tạo SQLite: {}", e.getMessage(), e);
        }
    }

    private void createTables() throws SQLException {
        String sql = """
            CREATE TABLE IF NOT EXISTS local_messages (
                id           INTEGER PRIMARY KEY AUTOINCREMENT,
                room_id      INTEGER NOT NULL,
                sender_id    INTEGER NOT NULL,
                sender_name  TEXT    NOT NULL,
                content      TEXT    NOT NULL,  -- plaintext (sau khi giải mã AES)
                message_type TEXT    DEFAULT 'TEXT',
                file_name    TEXT,
                timestamp    INTEGER NOT NULL,  -- epoch milliseconds
                is_sent      INTEGER DEFAULT 1  -- 1=sent by me, 0=received
            );
            CREATE INDEX IF NOT EXISTS idx_room_time
                ON local_messages(room_id, timestamp);
            """;
        try (Statement stmt = connection.createStatement()) {
            stmt.executeUpdate(sql);
        }
    }

    // ────────────────────────────────────────────────────────────
    // CRUD
    // ────────────────────────────────────────────────────────────

    /**
     * Lưu tin nhắn vào local history.
     */
    public void saveMessage(int roomId, int senderId, String senderName,
                             String content, String type, String fileName,
                             long timestamp, boolean isSent) {
        String sql = "INSERT INTO local_messages " +
                     "(room_id, sender_id, sender_name, content, message_type, file_name, timestamp, is_sent) " +
                     "VALUES (?, ?, ?, ?, ?, ?, ?, ?)";
        try (PreparedStatement ps = connection.prepareStatement(sql)) {
            ps.setInt(1, roomId);
            ps.setInt(2, senderId);
            ps.setString(3, senderName);
            ps.setString(4, content);
            ps.setString(5, type);
            ps.setString(6, fileName);
            ps.setLong(7, timestamp);
            ps.setInt(8, isSent ? 1 : 0);
            ps.executeUpdate();
        } catch (SQLException e) {
            log.error("Lỗi lưu local message: {}", e.getMessage());
        }
    }

    /**
     * Lấy lịch sử chat của một phòng (50 tin gần nhất).
     * @return danh sách [roomId, senderId, senderName, content, type, fileName, timestamp, isSent]
     */
    public List<Object[]> getHistory(int roomId, int limit) {
        String sql = "SELECT room_id, sender_id, sender_name, content, message_type, " +
                     "       file_name, timestamp, is_sent " +
                     "FROM local_messages WHERE room_id = ? " +
                     "ORDER BY timestamp DESC LIMIT ?";
        List<Object[]> results = new ArrayList<>();

        try (PreparedStatement ps = connection.prepareStatement(sql)) {
            ps.setInt(1, roomId);
            ps.setInt(2, limit);
            ResultSet rs = ps.executeQuery();

            while (rs.next()) {
                results.add(0, new Object[]{  // insert at 0 để reverse order
                    rs.getInt("room_id"),
                    rs.getInt("sender_id"),
                    rs.getString("sender_name"),
                    rs.getString("content"),
                    rs.getString("message_type"),
                    rs.getString("file_name"),
                    rs.getLong("timestamp"),
                    rs.getInt("is_sent") == 1
                });
            }
        } catch (SQLException e) {
            log.error("Lỗi lấy local history: {}", e.getMessage());
        }
        return results;
    }

    /**
     * Tìm kiếm tin nhắn theo từ khoá.
     */
    public List<Object[]> searchMessages(String keyword) {
        String sql = "SELECT room_id, sender_name, content, timestamp " +
                     "FROM local_messages WHERE content LIKE ? " +
                     "ORDER BY timestamp DESC LIMIT 50";
        List<Object[]> results = new ArrayList<>();

        try (PreparedStatement ps = connection.prepareStatement(sql)) {
            ps.setString(1, "%" + keyword + "%");
            ResultSet rs = ps.executeQuery();
            while (rs.next()) {
                results.add(new Object[]{
                    rs.getInt("room_id"),
                    rs.getString("sender_name"),
                    rs.getString("content"),
                    rs.getLong("timestamp")
                });
            }
        } catch (SQLException e) {
            log.error("Lỗi tìm kiếm: {}", e.getMessage());
        }
        return results;
    }

    /**
     * Xoá toàn bộ lịch sử của một phòng.
     */
    public void clearRoomHistory(int roomId) {
        String sql = "DELETE FROM local_messages WHERE room_id = ?";
        try (PreparedStatement ps = connection.prepareStatement(sql)) {
            ps.setInt(1, roomId);
            int deleted = ps.executeUpdate();
            log.info("Đã xoá {} tin nhắn local của room {}", deleted, roomId);
        } catch (SQLException e) {
            log.error("Lỗi xoá lịch sử: {}", e.getMessage());
        }
    }

    public void close() {
        try {
            if (connection != null && !connection.isClosed()) connection.close();
        } catch (SQLException e) {
            log.error("Lỗi đóng DB: {}", e.getMessage());
        }
    }
}
